airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -240,23 +274,21 @@ private void scanFile(MediaFile file, MusicFolder musicFolder, Instant lastScann
         indexManager.index(file);
 
         if (file.isDirectory()) {
-            for (MediaFile child : mediaFileService.getChildrenOf(file, true, false, false, false)) {
-                scanFile(child, musicFolder, lastScanned, albumCount, genres, isPodcast);
-            }
-            for (MediaFile child : mediaFileService.getChildrenOf(file, false, true, false, false)) {
-                scanFile(child, musicFolder, lastScanned, albumCount, genres, isPodcast);
-            }
+            ForkJoinTask.invokeAll(
+                    mediaFileService.getChildrenOf(file, true, true, false, false)
+                        .parallelStream()
+                        .map(child -> ForkJoinTask.adapt(() -> scanFile(child, musicFolder, lastScanned, albumCount, artists, albums, genres, isPodcast)))
+                        .collect(Collectors.toList()));
         } else {
             if (!isPodcast) {
-                updateAlbum(file, musicFolder, lastScanned, albumCount);
-                updateArtist(file, musicFolder, lastScanned, albumCount);
+                updateAlbum(file, musicFolder, lastScanned, albumCount, albums);
+                updateArtist(file, musicFolder, lastScanned, albumCount, artists);
             }
             statistics.incrementSongs(1);
         }
 
         updateGenres(file, genres);
         mediaFileDao.markPresent(file.getPath(), lastScanned);
-        artistDao.markPresent(file.getAlbumArtist(), lastScanned);
```

Comment:
in-memory model already holds the current status, no need to update db (only at the end)